arangod/RestHandler/RestDumpHandler.cpp

```diff
@@ -0,0 +1,269 @@
+constexpr double defaultTtl = 600.0;
+}  // namespace
+
+RestDumpHandler::RestDumpHandler(ArangodServer& server, GeneralRequest* request,
+                                 GeneralResponse* response)
+    : RestVocbaseBaseHandler(server, request, response) {}
+
+// main function that dispatches the different routes and commands
+RestStatus RestDumpHandler::execute() {
+  // we should not get here on coordinators, simply because of the
+  // request forwarding.
+  TRI_ASSERT(!ServerState::instance()->isCoordinator());
+
+  auto type = _request->requestType();
+  auto const& suffixes = _request->suffixes();
+  size_t const len = suffixes.size();
+
+  if (type == rest::RequestType::DELETE_REQ) {
+    if (len == 1) {
+      // end a dump
+      handleCommandDumpFinished();
+    } else {
+      generateError(
+          Result(TRI_ERROR_BAD_PARAMETER, "expecting DELETE /_api/dump/<id>"));
+    }
+  } else if (type == rest::RequestType::POST) {
+    if (len == 1 && suffixes[0] == "start") {
+      // start a dump
+      handleCommandDumpStart();
+    } else if (len == 2 && suffixes[0] == "next") {
+      // fetch next data from a dump
+      handleCommandDumpNext();
+    } else {
+      generateError(
+          Result(TRI_ERROR_BAD_PARAMETER,
+                 "expecting POST /_api/dump/start or /_api/dump/next/<id>"));
+    }
+  } else {
+    // invalid HTTP method
+    generateError(rest::ResponseCode::METHOD_NOT_ALLOWED,
+                  TRI_ERROR_HTTP_METHOD_NOT_ALLOWED);
+  }
+
+  return RestStatus::DONE;
+}
+
+RequestLane RestDumpHandler::lane() const {
+  auto type = _request->requestType();
+
+  if (type == rest::RequestType::DELETE_REQ) {
+    // delete should be prioritized, because it frees
+    // up resources
+    return RequestLane::CLUSTER_INTERNAL;
+  }
+  return RequestLane::SERVER_REPLICATION;
+}
+
+/// @brief returns the short id of the server which should handle this request
+ResultT<std::pair<std::string, bool>> RestDumpHandler::forwardingTarget() {
+  auto base = RestVocbaseBaseHandler::forwardingTarget();
+  if (base.ok() && !std::get<0>(base.get()).empty()) {
+    return base;
+  }
+
+  auto type = _request->requestType();
+  auto const& suffixes = _request->suffixes();
+  size_t const len = suffixes.size();
+  if (type == rest::RequestType::DELETE_REQ) {
+    if (len != 1) {
+      return ResultT<std::pair<std::string, bool>>::error(
+          TRI_ERROR_BAD_PARAMETER, "expecting DELETE /_api/dump/<id>");
+    }
+  } else if (type == rest::RequestType::POST) {
+    if ((len == 1 && suffixes[0] != "start") ||
+        (len == 2 && suffixes[0] != "next") || len < 1 || len > 2) {
+      return ResultT<std::pair<std::string, bool>>::error(
+          TRI_ERROR_BAD_PARAMETER,
+          "expecting POST /_api/dump/start or /_api/dump/next/<id>");
+    }
+  } else {
+    // invalid HTTP method
+    return ResultT<std::pair<std::string, bool>>::error(
+        TRI_ERROR_HTTP_METHOD_NOT_ALLOWED);
+  }
```

Comment:
This contains quite some duplication with `execute()`. Can we somehow avoid this?

Reply:
Pushed a change into the PR that should improve the situation, although I am still not very happy with it.